runtime/vm/visible.c

```diff
@@ -142,3 +155,55 @@ checkVisibility(J9VMThread *currentThread, J9Class* sourceClass, J9Class* destCl
 
 	return result;
 }
+
+#if defined(J9VM_OPT_VALHALLA_NESTMATES)
+static VMINLINE UDATA loadAndVerifyNestHost(J9VMThread *vmThread, J9Class *clazz, BOOLEAN isSourceClass) {
+	J9Class *nestHost = NULL;
+	UDATA result = J9_VISIBILITY_ALLOWED;
+	J9ROMClass *romClass = clazz->romClass;
+	J9UTF8 *nestHostName = J9ROMCLASS_NESTHOSTNAME(romClass);
+
+	/* If no nest host is named, class is own nest host */
+	if (NULL == nestHostName) {
+		nestHost = clazz;
+	} else {
+		UDATA classLoadingFlags = J9_FINDCLASS_FLAG_THROW_ON_FAIL;
+		nestHost = vmThread->javaVM->internalVMFunctions->internalFindClassUTF8(vmThread, J9UTF8_DATA(nestHostName), J9UTF8_LENGTH(nestHostName), clazz->classLoader, classLoadingFlags);
+
+		/* Nest host must be successfully loaded by the same classloader in the same package & verify the nest member */
+		if (NULL == nestHost) {
+			result = J9_VISIBILITY_NEST_HOST_LOADING_FAILURE_ERROR_SOURCE_CLASS;
+		} else if (clazz->packageID != nestHost->packageID) {
+			result = J9_VISIBILITY_NEST_HOST_DIFFERENT_PACKAGE_ERROR_SOURCE_CLASS;
+		} else {
+			/* The nest host must have a nestmembers attribute that claims this class. */
+			BOOLEAN verified = FALSE;
+			J9UTF8 *className = J9ROMCLASS_CLASSNAME(romClass);
+			J9SRP *nestMembers = J9ROMCLASS_NESTMEMBERS(nestHost->romClass);
+			U_16 nestMemberCount = nestHost->romClass->nestMemberCount;
+			U_16 i = 0;
+
+			for (i = 0; i < nestMemberCount; i++) {
+				J9UTF8 *nestMemberName = NNSRP_GET(nestMembers[i], J9UTF8*);
+				if (J9UTF8_EQUALS(className, nestMemberName)) {
+					verified = TRUE;
+					break;
+				}
+			}
+			if (!verified) {
+				result = J9_VISIBILITY_NEST_MEMBER_NOT_CLAIMED_ERROR_SOURCE_CLASS;
+			}
+		}
+
+		/* If any problem occurred in nest host verification, error code was set
+		 * assuming that the nest host is being loaded for the source class; if
+		 * not, add a -1 offset to change to dest class error code */
+		if ((J9_VISIBILITY_ALLOWED != result) && (TRUE != isSourceClass)) {
```

Comment:
Not a fan of this approach.  It assumes no one will ever renumber the constants and if they do, it will result in subtly wrong error messages.

I would rather see a new out parameter to indicate source / dest or a more obvious way of mapping between the two.  A table maybe?